test/core/end2end/tests/short_deadlines.c

```diff
@@ -0,0 +1,200 @@
+
+  cq_verifier *cqv = cq_verifier_create(f.cq);
+  grpc_op ops[6];
+  grpc_op *op;
+  grpc_metadata_array initial_metadata_recv;
+  grpc_metadata_array trailing_metadata_recv;
+  grpc_status_code status;
+  grpc_call_error error;
+  char *details = NULL;
+  size_t details_capacity = 0;
+
+  gpr_log(GPR_DEBUG, "test with %" PRIuPTR " ops, %d ms deadline", num_ops,
+          deadline_ms);
+
+  c = grpc_channel_create_call(
+      f.client, NULL, GRPC_PROPAGATE_DEFAULTS, f.cq, "/foo",
+      get_host_override_string("foo.test.google.fr:1234", config), deadline,
+      NULL);
+  GPR_ASSERT(c);
+
+  grpc_metadata_array_init(&initial_metadata_recv);
+  grpc_metadata_array_init(&trailing_metadata_recv);
+
+  memset(ops, 0, sizeof(ops));
+  op = ops;
+  op->op = GRPC_OP_RECV_STATUS_ON_CLIENT;
+  op->data.recv_status_on_client.trailing_metadata = &trailing_metadata_recv;
+  op->data.recv_status_on_client.status = &status;
+  op->data.recv_status_on_client.status_details = &details;
+  op->data.recv_status_on_client.status_details_capacity = &details_capacity;
+  op->flags = 0;
+  op->reserved = NULL;
+  op++;
+  op->op = GRPC_OP_RECV_INITIAL_METADATA;
+  op->data.recv_initial_metadata = &initial_metadata_recv;
+  op->flags = 0;
+  op->reserved = NULL;
+  op++;
+  op->op = GRPC_OP_SEND_INITIAL_METADATA;
+  op->data.send_initial_metadata.count = 0;
+  op->flags = 0;
+  op->reserved = NULL;
+  op++;
+  op->op = GRPC_OP_SEND_CLOSE_FROM_CLIENT;
+  op->flags = 0;
+  op->reserved = NULL;
+  op++;
+  GPR_ASSERT(num_ops <= (size_t)(op - ops));
+  error = grpc_call_start_batch(c, ops, num_ops, tag(1), NULL);
+  GPR_ASSERT(GRPC_CALL_OK == error);
+
+  CQ_EXPECT_COMPLETION(cqv, tag(1), 1);
+  cq_verify(cqv);
+
+  if (status != GRPC_STATUS_DEADLINE_EXCEEDED) {
+    gpr_log(GPR_ERROR,
+            "Expected GRPC_STATUS_DEADLINE_EXCEEDED (code %d), got code %d",
+            GRPC_STATUS_DEADLINE_EXCEEDED, status);
+    abort();
+  }
+
+  gpr_free(details);
+  grpc_metadata_array_destroy(&initial_metadata_recv);
+  grpc_metadata_array_destroy(&trailing_metadata_recv);
+
+  grpc_call_destroy(c);
+
+  cq_verifier_destroy(cqv);
+}
+
+static void test_invoke_short_deadline_request(grpc_end2end_test_config config,
+                                               size_t num_ops,
+                                               int deadline_ms) {
+  grpc_end2end_test_fixture f;
+
+  f = begin_test(config, __func__, NULL, NULL);
+  simple_request_body_with_deadline(config, f, num_ops, deadline_ms);
+  end_test(&f);
+  config.tear_down_data(&f);
+}
+
+void short_deadlines(grpc_end2end_test_config config) {
+  size_t i;
+  for (i = 1; i <= 4; i++) {
+    test_invoke_short_deadline_request(config, i, 0);
+    test_invoke_short_deadline_request(config, i, 1);
+    test_invoke_short_deadline_request(config, i, 5);
+    test_invoke_short_deadline_request(config, i, 10);
+    test_invoke_short_deadline_request(config, i, 15);
+    test_invoke_short_deadline_request(config, i, 30);
```

Comment:
Ah, very good point... I'll make some changes so that the server side delays long enough.